File: src/main/数组/t48/Rotate.java
package 数组.t48;

/**
 * @auther: sotowang
 * @date: 2019/12/04 20:12
 */
public class Rotate {
    public void rotate(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0] == null || matrix[0].length == 0) {
            return;
        }
        int rows = matrix.length;
        int cols = matrix[0].length;
        int start = 0;
        int end = rows - 1;
        //先将数组翻转
        while (start < end) {
            for (int i = 0; i < cols; i++) {
                int temp = matrix[start][i];
                matrix[start][i] = matrix[end][i];
                matrix[end][i] = temp;
            }
            start++;
            end--;
        }
        //再将数组没主对角线翻转
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < i; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
    }

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 2, 3},
                {4, 5, 6},
                {7, 8, 9},
        };
        new Rotate().rotate(matrix);
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                System.out.print(matrix[i][j] + ",");
            }
            System.out.println();
        }
    }
}

File: src/main/数组/t48/rotate.go
package t48

func rotate(matrix [][]int) {
	rows, cols := len(matrix), len(matrix[0])
	start, end := 0, cols-1
	for start < end {
		for i := 0; i < cols; i++ {
			matrix[start][i], matrix[end][i] = matrix[end][i], matrix[start][i]
		}
		start++
		end--
	}
	for i := 0; i < rows; i++ {
		for j := 0; j < i; j++ {
			matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
		}
	}
}
